### PR TITLE
Initial implementation of 'auto' for `min_resource`

### PR DESCRIPTION
**Pull request by crcrpar**
In this PR, I try to implement `'auto'` mode for `min_resource` in `SuccessiveHalvingPruner`.
The motivation is that I want to remove a hyperparameter in the hyperparameter optimization workflow.

The initial logic might be a bit too naive: if there is a `Trial` with its `state` is `COMPLETE`, set the `min_resource` to the hundredth of the number of steps of it.

<!--```
# master (commit e75c7f0db48978f9a77a3663d3b41a67305f54ed)
Study statistics:
  Number of finished trials:  100
  Number of pruned trials:  99
  Number of complete trials:  1
Best trial:
  Value:  0.9473684210526315
  Params:
    alpha: 0.001219699835809836

# this PR (commit f5e88dff51240f9d540a03732418161361e9f1ad)
Study statistics:
  Number of finished trials:  100
  Number of pruned trials:  96
  Number of complete trials:  4
Best trial:
  Value:  0.9736842105263158
  Params:
    alpha: 0.0023708809016833167
```-->

<!--```
diff --git a/examples/pruning/simple.py b/examples/pruning/simple.py
index b4446025..d0803a21 100644
--- a/examples/pruning/simple.py
+++ b/examples/pruning/simple.py
@@ -45,7 +45,8 @@ def objective(trial):


 if __name__ == '__main__':
-    study = optuna.create_study(direction='maximize')
+    pruner = optuna.pruners.SuccessiveHalvingPruner()
+    study = optuna.create_study(pruner=pruner, direction='maximize')
     study.optimize(objective, n_trials=100)

     pruned_trials = [t for t in study.trials if t.state == optuna.structs.TrialState.PRUNED]
```-->

### CODE REVIEW

File: optuna/pruners/successive_halving.py [hunk 1]
```diff
@@ -128,6 +146,25 @@ def prune(self, study, trial):
...
+        if not complete_trials:
+            return
+
+        # get the number of steps and divide it by 100.
+        complete_trial = complete_trials[0]  # type: FrozenTrial
+        last_step = complete_trial.last_step
+        if last_step is not None:
+            hundredth = last_step // 100
+            self._min_resource = hundredth if hundredth > 0 else 1
+            assert self._min_resource > 0
```
Comment (sile) [inline optuna/pruners/successive_halving.py]: It's obvious from the above line (L165) that this condition holds, so I think we can remove this assertion.

File: optuna/pruners/successive_halving.py [hunk 2]
```diff
@@ -103,9 +114,16 @@ def prune(self, study, trial):
 
         rung = _get_current_rung(trial)
         value = trial.intermediate_values[step]
-        trials = None
+        trials = None  # type: Optional[List[FrozenTrial]]
 
         while True:
+            if self._auto_min_resource:
```
Comment (sile) [inline optuna/pruners/successive_halving.py]: How about modifying this line as follows to reduce unnecessary `_estimate_in_resource()` calls once `_min_resource` is set?
```suggestion
            if self._auto_min_resource and self._min_resource is None:
```

Comment (sile) [reply optuna/pruners/successive_halving.py]: Furthermore, I think that the `_auto_min_resource` field can be removed from the class because it seems only `_min_resource` is needed to handle this condition.

File: optuna/pruners/successive_halving.py [hunk 3]
```diff
@@ -128,6 +146,25 @@ def prune(self, study, trial):
...
+
+        if trials is None:
+            return
+
+        complete_trials = [t for t in trials if t.state == TrialState.COMPLETE]
+
+        if not complete_trials:
+            return
+
+        # get the number of steps and divide it by 100.
```
Comment (sile) [inline optuna/pruners/successive_halving.py]:
```suggestion
        # Get the number of steps and divide it by 100.
```

File: tests/pruners_tests/test_successive_halving.py [hunk 4]
```diff
@@ -108,6 +108,35 @@ def test_successive_halving_pruner_with_nan():
...
+
+        for i in range(3):
+            trial.report(1.0 / (i + 1))
+            if trial.should_prune():
+                raise optuna.exceptions.TrialPruned()
+        return 1.0
+
+    study.optimize(objective, n_trials=3)
+    # FIXME (crcrpar)
+    # assert pruner._min_resource is not None and pruner._min_resource > 0
```
Comment (crcrpar) [inline tests/pruners_tests/test_successive_halving.py]: currently (https://github.com/optuna/optuna/pull/812/commits/c292d58a78042ec1122d78dd00292fd7dd8b46e6), this fails

File: optuna/pruners/successive_halving.py [hunk 5]
```diff
@@ -103,9 +119,17 @@ def prune(self, study, trial):
...
         rung = _get_current_rung(trial)
         value = trial.intermediate_values[step]
-        trials = None
+        trials = None  # type: Optional[List[FrozenTrial]]
 
         while True:
+            if self._min_resource is None:
+                if trials is None:
+                    trials = study.get_trials(deepcopy=False)
+                self._estimate_min_resource(trials)
```
Comment (hvy) [inline optuna/pruners/successive_halving.py]: How about making `_estimate_min_resource` a free function that return the value insetting of setting it to `self`? I think it improved readability.

File: optuna/pruners/successive_halving.py [hunk 6]
```diff
@@ -128,6 +152,24 @@ def prune(self, study, trial):
...
+            t.last_step for t in trials
+            if t.state == TrialState.COMPLETE and t.last_step is not None
+        ]
+
+        if not n_steps:
+            return
+
+        # Get the maximum number of steps and divide it by 100.
+        last_step = max(n_steps)
+        self._min_resource = max(int(last_step / 100), 1)
```
Comment (hvy) [inline optuna/pruners/successive_halving.py]: Can you simplify it as follows?
```suggestion
        self._min_resource = max(last_step // 100, 1)
```

File: tests/pruners_tests/test_successive_halving.py [hunk 7]
```diff
@@ -108,6 +108,34 @@ def test_successive_halving_pruner_with_nan():
...
+
+    pruner = optuna.pruners.SuccessiveHalvingPruner(min_resource='auto')
+    study = optuna.study.create_study(sampler=optuna.samplers.RandomSampler(), pruner=pruner)
+
+    assert pruner._min_resource is None
+
+    def objective(trial):
+        # type: (optuna.trial.Trial) -> float
+
+        for i in range(3):
```
Comment (hvy) [inline tests/pruners_tests/test_successive_halving.py]: How about parameterizing the number of steps to cover edge cases incl. 0 and 1?

File: optuna/pruners/successive_halving.py [hunk 8]
```diff
@@ -46,6 +49,9 @@ class SuccessiveHalvingPruner(BasePruner):
             A parameter for specifying the minimum resource allocated to a trial
             (in the `paper <http://arxiv.org/abs/1810.05934>`_ this parameter is
             referred to as :math:`r`).
+            This parameter defaults to ``'auto'`` that means this pruner estimates an appropriate
+            minimum resource by the required number of a ``COMPLETE``
+            :class:`~optuna.structs.FrozenTrial`'s steps.
```
Comment (hvy) [inline optuna/pruners/successive_halving.py]: I believe this descriptions is quite "internal" and could be a tiny bit more user friendly/descriptive. Something along.

> This parameter defaults to ``'auto'`` where the value is determined based on a heuristic that looks at the number of required steps for the first trial before completion.

**Comment by hvy** [comment]
Thanks for the PR. I was not aware of this heuristic. I left some comments if you could take a look.

File: optuna/pruners/successive_halving.py [hunk 9]
```diff
@@ -103,9 +119,17 @@ def prune(self, study, trial):
...
-        trials = None
+        trials = None  # type: Optional[List[FrozenTrial]]
 
         while True:
+            if self._min_resource is None:
+                if trials is None:
+                    trials = study.get_trials(deepcopy=False)
+                self._estimate_min_resource(trials)
+                if self._min_resource is None:
+                    return False
```
Comment (hvy) [inline optuna/pruners/successive_halving.py]: Should we keep returning `True` when the value is NaN, even for the first trial (the NaN check comes after this return)?

Comment (crcrpar) [reply optuna/pruners/successive_halving.py]: If return `True` here, the very first trial will be pruned. Is it acceptable?

Comment (hvy) [reply optuna/pruners/successive_halving.py]: Sorry, not sure I'm understanding you correctly but my point was to not return here at all but to delay this check/return to after the NaN check to keep the existing behavior.

File: tests/pruners_tests/test_successive_halving.py (hunk 7 above)
Comment (crcrpar) [reply tests/pruners_tests/test_successive_halving.py]: Sure thing.
I'll parametrize the number of reports and trials.

File: optuna/pruners/successive_halving.py (hunk 8 above)
Comment (crcrpar) [reply optuna/pruners/successive_halving.py]: Positive.

I'll use your description.

File: optuna/pruners/successive_halving.py (hunk 5 above)
Comment (crcrpar) [reply optuna/pruners/successive_halving.py]: Ah, you're right. This conceals that `self._estimate_min_resource` modifies `self._min_resource`.

File: optuna/pruners/successive_halving.py (hunk 9 above)
Comment (crcrpar) [reply optuna/pruners/successive_halving.py]: > the NaN check comes after this return

The check exists there for passing mypy check.

Comment (crcrpar) [reply optuna/pruners/successive_halving.py]: > Should we keep returning `True` when the value is NaN, even for the first trial (the NaN check comes after this return)?

This sounded like L.130 should return `True` instead of `False` to me...

Comment (crcrpar) [reply optuna/pruners/successive_halving.py]: > Should we keep returning `True` when the value is NaN, even for the first trial (the NaN check comes after this return)?

This sounded like L.130 should return `True` instead of `False` to me...

Comment (hvy) [reply optuna/pruners/successive_halving.py]: Sorry, it's probably correct the way it is, returning `False` here. (The rung promotion step below cannot be computed unless the minimum resource is determined.)

Comment (crcrpar) [reply optuna/pruners/successive_halving.py]: > Sorry, not sure I'm understanding you correctly but my point was to not return here at all but to delay this check/return to after the NaN check to keep the existing behavior.

```python
def prune(self, study: Study, trial: FrozenTrial) -> bool:

    ...

    while True:
        if self._min_resource is None:
            if trials is None: trials = study.get_trials(False)
            self._min_resource = _estimate_min_resource(trials)

        if self._min_resource is None:
            return False
        else:
            ...
```

What you assume is like the above snippet?
I want to see what you think better in a snippet as I don't still get the point.

Comment (hvy) [reply optuna/pruners/successive_halving.py]: Not really, that seems identical to the current implementation. As I wrote earlier,  https://github.com/optuna/optuna/pull/812/files#r364126282 however, you can ignore this comment as the current implementation looks correct.

**Comment by hvy** [comment]
LGTM!
